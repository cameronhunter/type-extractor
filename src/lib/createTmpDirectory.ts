import { mkdtemp } from 'node:fs/promises';
import { tmpdir } from 'node:os';
import { join } from 'node:path';

export async function createTmpDirectory(): Promise<string> {
    return mkdtemp(join(tmpdir(), 'type-extractor-'));
}
