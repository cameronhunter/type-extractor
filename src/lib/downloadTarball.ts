import { ok } from 'assert';
import { createWriteStream } from 'fs';
import { mkdtemp } from 'fs/promises';
import { tmpdir } from 'os';
import * as path from 'path';
import { Readable } from 'stream';
import { finished } from 'stream/promises';
import { fetch } from 'undici';
import type { URL } from 'url';

/**
 * Find and download a tarball from the registry.
 *
 * @returns the path to the downloaded tarball.
 */
export async function downloadTarball(url: URL): Promise<string> {
    const tmpDirectory = await mkdtemp(path.join(tmpdir(), `type-extractor-tarball`));
    const destinationFilepath = path.join(tmpDirectory, path.basename(url.pathname));

    const response = await fetch(url);

    ok(response.ok, `Response failed with status ${response.status}: ${response.statusText}`);

    await finished(Readable.fromWeb(response.body!).pipe(createWriteStream(destinationFilepath, { flags: 'wx' })));

    return destinationFilepath;
}
