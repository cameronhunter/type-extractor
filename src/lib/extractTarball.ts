import { mkdtemp } from 'fs/promises';
import * as path from 'path';
import { tmpdir } from 'os';
import { extract } from 'tar';
import { minimatch } from 'minimatch';

export async function extractTarball(file: string, filter: string[] = []): Promise<string> {
    const tmpDirectory = await mkdtemp(path.join(tmpdir(), 'type-extractor-package'));

    // Files in the tarball all in a top-level `package` directory that we don't want.
    const strip = 1;

    await extract({
        file,
        cwd: tmpDirectory,
        strip,
        filter(filePath) {
            filePath = filePath.split(path.sep).slice(strip).join(path.sep);
            return filter.length === 0 || filter.some((filter) => minimatch(filePath, filter));
        },
    });

    return tmpDirectory;
}
