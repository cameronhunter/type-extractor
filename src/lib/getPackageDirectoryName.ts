/**
 * Creating `@types` packages for scoped packages requires a special naming
 * convention of `@types/scope__package`.
 */
export function getPackageDirectoryName(packageName: string): string {
    const [_, scope, name] = packageName.match(/(?:@(.+)\/)(.+)/) || [undefined, undefined, packageName];
    return scope ? `${scope}__${name}` : name;
}
