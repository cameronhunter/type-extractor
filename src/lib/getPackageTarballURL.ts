import { ok } from 'node:assert';
import { fetch } from 'undici';
import { URL } from 'node:url';

export interface PackageDetails {
    'dist-tags': { [tag: string]: string };
    versions: { [version: string]: { dist: { tarball: string } } };
}

/**
 * Use the npm registry to find the tarball URL for a particular version of a
 * package.
 *
 * @returns the URL to the package tarball.
 */
export async function getPackageTarballURL(packageName: string, version: string = 'latest'): Promise<URL> {
    const response = await fetch(`https://registry.npmjs.org/${packageName}`);

    ok(response.ok, `Response failed with status ${response.status}: ${response.statusText}`);

    const packageDetails = (await response.json()) as PackageDetails;

    if (version in packageDetails['dist-tags']) {
        const distTagVersion = packageDetails['dist-tags'][version];
        ok(distTagVersion, `Expected a ${version} dist-tag version to exist.`);
        version = distTagVersion;
    }

    const tarball = packageDetails.versions[version]?.dist.tarball;

    ok(tarball, `Expected a tarball to exist for version "${version}".`);

    return new URL(tarball);
}
