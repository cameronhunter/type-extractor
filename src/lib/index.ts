export function getPackageURL() {
  return null;
}
