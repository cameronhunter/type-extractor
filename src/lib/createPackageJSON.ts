import { getPackageDirectoryName } from './getPackageDirectoryName';
import depcheck, { PackageDependencies } from 'depcheck';

export interface PackageJSON {
    name: string;
    exports?: { [path: string]: string | { types?: string } } | undefined;
    dependencies?: PackageDependencies | undefined;
    [other: string]: unknown;
}

export async function createTypesPackageJSON(packageDirectory: string, packageJSON: PackageJSON): Promise<PackageJSON> {
    return {
        ...packageJSON,
        name: rewriteName(packageJSON.name),
        exports: rewriteExports(packageJSON.exports),
        dependencies: await rewriteDependencies(packageDirectory, packageJSON['dependencies']),

        /**
         * Remove these properties, as they have nothing to do with types.
         */
        bin: undefined,
        devDependencies: undefined,
        engines: undefined,
        files: undefined,
        main: undefined,
        optionalDependencies: undefined,
        peerDependencies: undefined,
        peerDependenciesMeta: undefined,
        repository: undefined,
        scripts: undefined,

        /**
         * Remove a few random properties
         */
        husky: undefined,
        jest: undefined,
        prettier: undefined,
        wireit: undefined,
    };
}

export function rewriteName(name: string): string {
    return `@types/${getPackageDirectoryName(name)}`;
}

/**
 * Rewrite package exports to filter out anything that isn't type-related.
 */
export function rewriteExports(packageExports: PackageJSON['exports'] | undefined): PackageJSON['exports'] | undefined {
    if (!packageExports) {
        return undefined;
    }

    return Object.entries(packageExports).reduce((state, [key, value]) => {
        if (typeof value === 'string') {
            if (key.endsWith('.d.ts') || value.endsWith('.d.ts')) {
                state[key] = value;
            }
        } else {
            if (value.types) {
                state[key] = { types: value.types };
            }
        }

        return state;
    }, {} as Exclude<PackageJSON['exports'], undefined>);
}

/**
 * Remove dependencies that aren't used in the type definitions.
 */
async function rewriteDependencies(
    packageDirectory: string,
    dependencies: PackageDependencies | undefined
): Promise<PackageDependencies | undefined> {
    if (!dependencies) {
        return undefined;
    }

    const { using } = await depcheck(packageDirectory, {
        ignoreBinPackage: true,
        skipMissing: true,
        parsers: {
            '**/*.d.ts': depcheck.parser.typescript,
        },
        detectors: [depcheck.detector.importDeclaration],
        package: {
            dependencies,
        },
    });

    return Object.keys(using).reduce((state, dependency) => {
        state[dependency] = dependencies[dependency]!;
        return state;
    }, {} as PackageDependencies);
}
