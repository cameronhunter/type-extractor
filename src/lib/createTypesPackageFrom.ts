import { doesNotMatch, ok } from 'node:assert';
import { mkdtemp, rename, writeFile } from 'node:fs/promises';
import * as path from 'node:path';
import { getPackageDirectoryName } from './getPackageDirectoryName';
import { getPackageTarballURL } from './getPackageTarballURL';
import { downloadTarball } from './downloadTarball';
import { extractTarball } from './extractTarball';
import { tmpdir } from 'node:os';
import { createTypesPackageJSON } from './createPackageJSON';
import { existsSync } from 'node:fs';

export async function createTypesPackageFrom(packageName: string, version?: string): Promise<string> {
    doesNotMatch(packageName, /^@types\/.+/, `There's no need to extract types from a DefinitelyTyped package.`);

    const url = await getPackageTarballURL(packageName, version);
    const tarball = await downloadTarball(url);
    const packageDirectory = await extractTarball(tarball, ['package.json', '**/*.d.ts', 'LICENSE*']);

    const packageJSONPath = path.join(packageDirectory, 'package.json');
    const packageJSON = require(packageJSONPath);

    ok(
        existsSync(path.resolve(packageDirectory, packageJSON.types || 'index.d.ts')),
        `Expected type definitions to exist in ${packageName}@${packageJSON.version}.`
    );

    const newPackageName = getPackageDirectoryName(packageName);
    const newPackageJSON = await createTypesPackageJSON(packageDirectory, packageJSON);

    await writeFile(packageJSONPath, JSON.stringify(newPackageJSON, null, 2), 'utf-8');

    await writeFile(
        path.join(packageDirectory, 'README.md'),
        `# ${newPackageName}\n\nTypescript type definitions extracted from \`${packageName}@${packageJSON.version}\`.\n`,
        'utf-8'
    );

    const destinationDirectory = path.join(await mkdtemp(tmpdir()), newPackageName);

    await rename(packageDirectory, destinationDirectory);

    return destinationDirectory;
}
