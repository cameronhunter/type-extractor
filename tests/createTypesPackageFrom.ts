import { test, expect } from 'vitest';
import { createTypesPackageFrom } from '../src/lib/createTypesPackageFrom';
import isCI from 'is-ci';

test('Throws an error when using @types packages', async () => {
    await expect(createTypesPackageFrom('@types/puppeteer')).rejects.toThrowErrorMatchingInlineSnapshot(
        '"There\'s no need to extract types from a DefinitelyTyped package."'
    );
});

test.skipIf(isCI)('it creates a @types package', async () => {
    const pkg = await createTypesPackageFrom('puppeteer-core', '20.7.4');

    const packageJSON = require(`${pkg}/package.json`);

    expect(packageJSON).toMatchSnapshot();
});
