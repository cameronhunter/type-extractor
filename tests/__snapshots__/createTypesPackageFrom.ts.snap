// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`it creates a @types package 1`] = `
{
  "author": "The Chromium Authors",
  "dependencies": {
    "@puppeteer/browsers": "1.4.3",
    "chromium-bidi": "0.4.16",
    "devtools-protocol": "0.0.1135028",
    "ws": "8.13.0",
  },
  "description": "A high-level API to control headless Chrome over the DevTools Protocol",
  "exports": {
    ".": {
      "types": "./lib/types.d.ts",
    },
  },
  "keywords": [
    "puppeteer",
    "chrome",
    "headless",
    "automation",
  ],
  "license": "Apache-2.0",
  "name": "@types/puppeteer-core",
  "type": "commonjs",
  "types": "./lib/types.d.ts",
  "version": "20.7.4",
}
`;
